File: Chapter 17/movie_plugin/android/src/main/kotlin/com/bpb/movie_plugin/MoviePlugin.kt
package com.bpb.movie_plugin


import android.content.Context
import android.content.SharedPreferences
import android.util.Log
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

enum class MethodCalls(val value: String) {
    GetString("getString"),
    SetString("setString"),
    Clear("clear"),
    ContainsKey("containsKey"),
    GetInt("getInt"),
    SetInt("setInt"),
    GetDouble("getDouble"),
    SetDouble("setDouble"),
    GetBool("getBool"),
    SetBool("setBool"),
}

/** MoviePlugin */
class MoviePlugin : FlutterPlugin, MethodCallHandler {
    /// The MethodChannel that will the communication between Flutter and native Android
    ///
    /// This local reference serves to register the plugin with the Flutter Engine and unregister it
    /// when the Flutter Engine is detached from the Activity
    private lateinit var channel: MethodChannel
    private lateinit var context: Context
    private lateinit var sharedPreferences: SharedPreferences

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "movie_plugin")
        channel.setMethodCallHandler(this)
        context = flutterPluginBinding.applicationContext
        sharedPreferences = context.getSharedPreferences("movie_plugin", Context.MODE_PRIVATE)
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        Log.e("MoviePlugin", "onMethodCall: ${call.method}")
        when (call.method) {
            MethodCalls.GetString.value -> {
                val key = call.argument<String>("key")
                Log.e("MoviePlugin", "key: $key")
                if (key == null) {
                    result.error("INVALID_ARGUMENT", "Key cannot be null", null)
                    return
                }
                getString(key, result)
            }

            MethodCalls.SetString.value -> {
                val key = call.argument<String>("key")
                val value = call.argument<String>("value")
                Log.e("MoviePlugin", "key: $key value: $value")
                if (key == null) {
                    result.error("INVALID_ARGUMENT", "Key cannot be null", null)
                    return
                }
                if (value == null) {
                    result.error("INVALID_ARGUMENT", "value cannot be null", null)
                    return
                }
                setString(key, value, result)
            }

            MethodCalls.Clear.value -> {
                clear(result)
            }

            MethodCalls.ContainsKey.value -> {
                val key = call.argument<String>("key")
                if (key == null) {
                    result.error("INVALID_ARGUMENT", "Key cannot be null", null)
                    return
                }
                containsKey(key, result)
            }

            MethodCalls.GetInt.value -> {
                val key = call.argument<String>("key")
                if (key == null) {
                    result.error("INVALID_ARGUMENT", "Key cannot be null", null)
                    return
                }
                getInt(key, result)
            }

            MethodCalls.SetInt.value -> {
                val key = call.argument<String>("key")
                val value = call.argument<Int>("value")
                if (key == null) {
                    result.error("INVALID_ARGUMENT", "Key cannot be null", null)
                    return
                }
                if (value == null) {
                    result.error("INVALID_ARGUMENT", "value cannot be null", null)
                    return
                }
                setInt(key, value, result)
            }

            MethodCalls.GetDouble.value -> {
                val key = call.argument<String>("key")
                if (key == null) {
                    result.error("INVALID_ARGUMENT", "Key cannot be null", null)
                    return
                }
                getDouble(key, result)
            }

            MethodCalls.SetDouble.value -> {
                val key = call.argument<String>("key")
                val value = call.argument<Double>("value")
                if (key == null) {
                    result.error("INVALID_ARGUMENT", "Key cannot be null", null)
                    return
                }
                if (value == null) {
                    result.error("INVALID_ARGUMENT", "value cannot be null", null)
                    return
                }
                setDouble(key, value, result)
            }

            MethodCalls.GetBool.value -> {
                val key = call.argument<String>("key")
                if (key == null) {
                    result.error("INVALID_ARGUMENT", "Key cannot be null", null)
                    return
                }
                getBool(key, result)
            }

            MethodCalls.SetBool.value -> {
                val key = call.argument<String>("key")
                val value = call.argument<Boolean>("value")
                if (key == null) {
                    result.error("INVALID_ARGUMENT", "Key cannot be null", null)
                    return
                }
                if (value == null) {
                    result.error("INVALID_ARGUMENT", "value cannot be null", null)
                    return
                }
                setBool(key, value, result)
            }

        }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }

    private fun getString(key: String, result: Result) {
        val value = sharedPreferences.getString(key, null)
        Log.e("MoviePlugin", "getString: key: $key value: $value")
        result.success(value)
    }

    private fun setString(key: String, value: String, result: Result) {
        val commitResult = sharedPreferences.edit().putString(key, value).commit()
        Log.e("MoviePlugin", "setString: key: $key value: $value commitResult: $commitResult")
        result.success(commitResult)
    }

    private fun clear(result: Result) {
        val commitResult = sharedPreferences.edit().clear().commit()
        result.success(commitResult)
    }

    private fun containsKey(key: String, result: Result) {
        val value = sharedPreferences.contains(key)
        result.success(value)
    }

    private fun getInt(key: String, result: Result) {
        val value = sharedPreferences.getInt(key, 0)
        result.success(value)
    }

    private fun setInt(key: String, value: Int, result: Result) {
        val commitResult = sharedPreferences.edit().putInt(key, value).commit()
        result.success(commitResult)
    }

    private fun getDouble(key: String, result: Result) {
        val value = sharedPreferences.getFloat(key, 0.0f).toDouble()
        result.success(value)
    }

    private fun setDouble(key: String, value: Double, result: Result) {
        val commitResult = sharedPreferences.edit().putFloat(key, value.toFloat()).commit()
        result.success(commitResult)
    }

    private fun getBool(key: String, result: Result) {
        val value = sharedPreferences.getBoolean(key, false)
        result.success(value)
    }

    private fun setBool(key: String, value: Boolean, result: Result) {
        val commitResult = sharedPreferences.edit().putBoolean(key, value).commit()
        result.success(commitResult)
    }


}

File: Chapter 18/movie_plugin/ios/Classes/MoviePlugin.swift
import Flutter
import UIKit

enum MethodCalls: String {
  case getString = "getString"
  case setString = "setString"
  case clear = "clear"
  case containsKey = "containsKey"
  case getInt = "getInt"
  case setInt = "setInt"
  case getBool = "getBool"
  case setBool = "setBool"
  case getDouble = "getDouble"
  case setDouble = "setDouble"
}

public class MoviePlugin: NSObject, FlutterPlugin {
  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(name: "movie_plugin", binaryMessenger: registrar.messenger())
    let instance = MoviePlugin()
    registrar.addMethodCallDelegate(instance, channel: channel)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    let args = call.arguments as? [String: Any]
    switch MethodCalls(rawValue: call.method) {
    case .getString:
      guard let key = args?["key"] as? String else {
        result(false)
        return
      }
      getString(key: key, result: result)

    case .setString:
      guard let key = args?["key"] as? String, let value = args?["value"] as? String else {
        result(false)
        return
      }
      setString(key: key, value: value, result: result)

    case .clear:
      clear()
      result(true)

    case .containsKey:
      guard let key = args?["key"] as? String else {
        result(false)
        break
      }
      containsKey(key: key, result: result)

    case .getInt:
      guard let key = args?["key"] as? String else {
        result(false)
        return
      }
      getInt(key: key, result: result)

    case .setInt:
      guard let key = args?["key"] as? String, let value = args?["value"] as? Int else {
        result(false)
        return
      }
      setInt(key: key, value: value, result: result)

    case .getBool:
      guard let key = args?["key"] as? String else {
        result(false)
        return
      }
      getBool(key: key, result: result)

    case .setBool:
      guard let key = args?["key"] as? String, let value = args?["value"] as? Bool else {
        result(false)
        return
      }
      setBool(key: key, value: value, result: result)

    case .getDouble:
      guard let key = args?["key"] as? String else {
        result(false)
        return
      }
      getDouble(key: key, result: result)

    case .setDouble:
      guard let key = args?["key"] as? String, let value = args?["value"] as? Double else {
        result(false)
        return
      }
      setDouble(key: key, value: value, result: result)

    default:
      result(FlutterMethodNotImplemented)
    }
  }

  func getString(key: String, result: @escaping FlutterResult) {
    let defaults = UserDefaults.standard
    result(defaults.string(forKey: key))
  }

  func setString(key: String, value: String, result: @escaping FlutterResult) {
    let defaults = UserDefaults.standard
    defaults.set(value, forKey: key)
    result(true)
  }

  func clear() {
    let defaults = UserDefaults.standard
    if let appDomain = Bundle.main.bundleIdentifier,
      let prefs = defaults.persistentDomain(forName: appDomain)
    {
      for (key, _) in prefs {
        defaults.removeObject(forKey: key)
      }
    }
  }

  func containsKey(key: String, result: @escaping FlutterResult) {
    let defaults = UserDefaults.standard
    let value = defaults.object(forKey: key)
    result(value != nil)
  }

  func getInt(key: String, result: @escaping FlutterResult) {
    let defaults = UserDefaults.standard
    result(defaults.integer(forKey: key))
  }

  func setInt(key: String, value: Int, result: @escaping FlutterResult) {
    let defaults = UserDefaults.standard
    defaults.set(value, forKey: key)
    result(true)
  }

  func getBool(key: String, result: @escaping FlutterResult) {
    let defaults = UserDefaults.standard
    result(defaults.bool(forKey: key))
  }

  func setBool(key: String, value: Bool, result: @escaping FlutterResult) {
    let defaults = UserDefaults.standard
    defaults.set(value, forKey: key)
    result(true)
  }

  func getDouble(key: String, result: @escaping FlutterResult) {
    let defaults = UserDefaults.standard
    result(defaults.double(forKey: key))
  }

  func setDouble(key: String, value: Double, result: @escaping FlutterResult) {
    let defaults = UserDefaults.standard
    defaults.set(value, forKey: key)
    result(true)
  }

}
